In [2]:
import pandas as pd

df = pd.read_csv('ipl.csv', low_memory=False)   

In [3]:
df.shape

(283678, 65)

In [4]:
df.head()

,Unnamed: 0,match_id,date,match_type,event_name,innings,batting_team,bowling_team,over,ball,...,team_balls,team_wicket,new_batter,power_surge_start,batter_runs,batter_balls,bowler_wicket,batting_partners,next_batter,striker_out
0,141607,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,1,...,1,0,NaN,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
1,141608,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,2,...,2,0,NaN,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
2,141609,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,...,2,0,NaN,NaN,0,1,0,"('BB McCullum', 'SC Ganguly')",NaN,False
3,141610,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,3,...,3,0,NaN,NaN,0,2,0,"('BB McCullum', 'SC Ganguly')",NaN,False
4,141611,335982,2008-04-18,T20,Indian Premier League,1,Kolkata Knight Riders,Royal Challengers Bangalore,0,4,...,4,0,NaN,NaN,0,3,0,"('BB McCullum', 'SC Ganguly')",NaN,False


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 283678 entries, 0 to 283677
Data columns (total 65 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         283678 non-null  int64  
 1   match_id           283678 non-null  int64  
 2   date               283678 non-null  object 
 3   match_type         283678 non-null  object 
 4   event_name         283678 non-null  object 
 5   innings            283678 non-null  int64  
 6   batting_team       283678 non-null  object 
 7   bowling_team       283678 non-null  object 
 8   over               283678 non-null  int64  
 9   ball               283678 non-null  int64  
 10  ball_no            283678 non-null  float64
 11  batter             283678 non-null  object 
 12  bat_pos            283678 non-null  int64  
 13  runs_batter        283678 non-null  int64  
 14  balls_faced        283678 non-null  int64  
 15  bowler             283678 non-null  object 
 16  va

In [6]:
df.describe()

,Unnamed: 0,match_id,innings,over,ball,ball_no,bat_pos,runs_batter,balls_faced,valid_ball,...,year,balls_per_over,overs,team_runs,team_balls,team_wicket,power_surge_start,batter_runs,batter_balls,bowler_wicket
count,283678.000000,2.836780e+05,283678.000000,283678.000000,283678.000000,283678.000000,283678.000000,283678.000000,283678.000000,283678.000000,...,283678.000000,283678.0,283678.0,283678.000000,283678.000000,283678.000000,0.0,283678.000000,283678.000000,283678.000000
mean,141838.500000,9.535684e+05,1.482787,9.190833,3.488847,9.539717,3.616445,1.281421,0.967082,0.962912,...,2016.889406,6.0,20.0,77.423099,58.596328,2.460624,NaN,18.376621,13.993813,0.045530
std,81890.929169,3.865091e+05,0.502511,5.680845,1.708286,5.682271,2.170569,1.655126,0.178421,0.188977,...,5.352468,0.0,0.0,50.167985,34.113517,2.102515,NaN,18.607095,11.815576,0.208465
min,0.000000,3.359820e+05,1.000000,0.000000,1.000000,0.100000,1.000000,0.000000,0.000000,0.000000,...,2008.000000,6.0,20.0,0.000000,0.000000,0.000000,NaN,0.000000,0.000000,0.000000
25%,70919.250000,5.483590e+05,1.000000,4.000000,2.000000,4.500000,2.000000,0.000000,1.000000,1.000000,...,2012.000000,6.0,20.0,36.000000,29.000000,1.000000,NaN,4.000000,5.000000,0.000000
50%,141838.500000,1.082612e+06,1.000000,9.000000,3.000000,9.400000,3.000000,1.000000,1.000000,1.000000,...,2017.000000,6.0,20.0,73.000000,58.000000,2.000000,NaN,12.000000,11.000000,0.000000
75%,212757.750000,1.304066e+06,2.000000,14.000000,5.000000,14.400000,5.000000,1.000000,1.000000,1.000000,...,2022.000000,6.0,20.0,114.000000,88.000000,4.000000,NaN,27.000000,20.000000,0.000000
max,283677.000000,1.529267e+06,6.000000,19.000000,7.000000,19.600000,11.000000,6.000000,1.000000,1.000000,...,2026.000000,6.0,20.0,287.000000,121.000000,10.000000,NaN,175.000000,73.000000,1.000000


In [7]:
df.isnull().sum()

Unnamed: 0               0
match_id                 0
date                     0
match_type               0
event_name               0
                     ...  
batter_balls             0
bowler_wicket            0
batting_partners         0
next_batter         270085
striker_out              0
Length: 65, dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.columns.tolist()

['Unnamed: 0',
 'match_id',
 'date',
 'match_type',
 'event_name',
 'innings',
 'batting_team',
 'bowling_team',
 'over',
 'ball',
 'ball_no',
 'batter',
 'bat_pos',
 'runs_batter',
 'balls_faced',
 'bowler',
 'valid_ball',
 'runs_extras',
 'runs_total',
 'runs_bowler',
 'runs_not_boundary',
 'extra_type',
 'non_striker',
 'non_striker_pos',
 'wicket_kind',
 'player_out',
 'fielders',
 'runs_target',
 'review_batter',
 'team_reviewed',
 'review_decision',
 'umpire',
 'umpires_call',
 'player_of_match',
 'match_won_by',
 'win_outcome',
 'toss_winner',
 'toss_decision',
 'venue',
 'city',
 'day',
 'month',
 'year',
 'season',
 'gender',
 'team_type',
 'superover_winner',
 'result_type',
 'method',
 'balls_per_over',
 'overs',
 'event_match_no',
 'stage',
 'match_number',
 'team_runs',
 'team_balls',
 'team_wicket',
 'new_batter',
 'power_surge_start',
 'batter_runs',
 'batter_balls',
 'bowler_wicket',
 'batting_partners',
 'next_batter',
 'striker_out']

## Dataset Overview

- Ball-by-ball IPL data, 2008–2025 — 283,678 rows x 65 columns
- Granularity: one row per ball bowled
- No duplicate rows
- `next_batter` is mostly null by design (only filled when a wicket falls) — not a data quality issue
- `Unnamed: 0` is a junk index column, to be dropped later
- Other missing values likely concentrated in wicket-related and DRS-related columns, since those only apply when relevant events happen

## Questions to Explore

1. Does winning the toss actually correlate with winning the match, and does that vary by venue?
2. Which teams/players perform best in run-chases vs defending totals?
3. How has scoring rate (runs per over) evolved across seasons — has the game gotten more aggressive over time?

In [10]:
df['toss_winner_won'] = df['toss_winner'] == df['match_won_by']
df['toss_winner_won'].value_counts(normalize=True)

toss_winner_won
True     0.505302
False    0.494698
Name: proportion, dtype: float64

Winning the toss provided almost no advantage, toss winners won 50.5% of matches, essentially identical to random chance (50%). This suggests the toss itself is not a meaningful predictor of match outcome across the full dataset.

In [12]:
df.groupby('venue')['match_id'].nunique().sort_values(ascending=False).head(10)

venue
Eden Gardens                                 77
Wankhede Stadium                             73
M Chinnaswamy Stadium                        65
Feroz Shah Kotla                             60
Wankhede Stadium, Mumbai                     55
Rajiv Gandhi International Stadium, Uppal    49
MA Chidambaram Stadium, Chepauk              48
Sawai Mansingh Stadium                       47
Dubai International Cricket Stadium          46
MA Chidambaram Stadium, Chepauk, Chennai     37
Name: match_id, dtype: int64

A few venues appeared to show 100% toss-win correlation, but this is a small-sample artifact, venues with only 1-2 matches show extreme percentages by chance. Venues with substantial match history stay close to the overall 50.5% baseline, confirming that toss provides no meaningful advantage anywhere in this dataset.